std/range/primitives.d

```diff
@@ -2004,11 +2004,12 @@ ElementType!R moveAt(R)(R r, size_t i)
 
 /**
 Implements the range interface primitive $(D empty) for built-in
-arrays. Due to the fact that nonmember functions can be called with
+arrays and associative arrays. Due to the fact that nonmember functions can be called with
 the first argument using the dot notation, $(D array.empty) is
 equivalent to $(D empty(array)).
  */
-@property bool empty(T)(in T[] a) @safe pure nothrow @nogc
+@property bool empty(T)(auto ref scope const(T) a) @safe pure nothrow @nogc
```

Comment:
please drop the const here

Reply:
If `const` is removed we'll have different template instances for each mutability class (up to 3 for each distinct `Unqual!T`. Also we users will see the following warnings:
```D
void main()
{    
    S ms;
    assert(!ms.empty);
    
    const S cs;
    assert(!cs.empty);

    immutable S ims;
    assert(!ims.empty);
}

struct S
{
    @disable this(this);
    @disable void opAssign(const ref S other);
    size_t length = 15;
}

import std.range.primitives : hasLength;
import std.traits : isNarrowString;

bool empty(T)(auto ref T a)
if (hasLength!T || isNarrowString!T)
{
    pragma (msg, __FUNCTION__);    
    return !a.length;
}
```

```
onlineapp.empty!(S).empty
/dlang/dmd/linux/bin64/../../src/phobos/std/range/primitives.d(1411): Note: length must have type size_t on all systems, please update your code by December 2017.
onlineapp.empty!(const(S)).empty
/dlang/dmd/linux/bin64/../../src/phobos/std/range/primitives.d(1411): Note: length must have type size_t on all systems, please update your code by December 2017.
onlineapp.empty!(immutable(S)).empty
```

Reply:
@RazvanN7 please drop `@safe pure nothrow @nogc` as this enforces that if `T.length` is a function it must have all of those attributes, which limits the generality. As a rule of thumb, template functions should not have explicit attributes and  should instead rely on attribute inference.